### PR TITLE
Initial version - foreman namespace/template

### PR DESCRIPTION
**Pull request by lzap**
What's up! This is an initial version of foreman namespace. It currently does generate JSON files but foreman.yml is ignored for some reason.

```
$ cd templates/foreman
$ make
$ grep foreman *json
org.foreman.viaq-cdm.2.4.4.template.json:  "template": "project.foreman-*"
org.foreman.viaq-cdm.5.5.2.template.json:  "template": "project.foreman-*"
```

Seeking for advice, thanks.

### CODE REVIEW

**Comment by lzap** [comment]
Fixed YAML syntax. YAML is hard! Ready for review, not yet ready for merge - still WIP

File: templates/foreman/template.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+skeleton_path: ../skeleton.json
+skeleton_index_pattern_path: ../skeleton-index-pattern.json
+
+elasticsearch_template:
+  name: org.foreman.viaq-cdm
+  index_pattern: "project.foreman-*"
```
Comment (richm) [inline templates/foreman/template.yml:6]: What is your intended index name format?  e.g. `project.foreman-$clustername.*`  That is - what comes after the `foreman-`?

Comment (lzap) [reply templates/foreman/template.yml:6]: Foreman/Satellite has no concept of clusters, it's always a single (HA) deployment. I copy-pasted this and there are no intentions. I struggle to find a better index name or pattern.

Comment (richm) [reply templates/foreman/template.yml:6]: In the ovirt case, they use `project.ovirt-metrics-*` for their metrics/monitoring data, and `project.ovirt-logs-*` for their logging data.  The `*` part is for the node type e.g. `ovirt-metrics-engine` is for the ovirt engine - perhaps you could do something similar for foreman - note that the part after `project` corresponds to an OpenShift project/namespace name - so for example in the ovirt case, they have to create `ovirt-metrics-engine` and `ovirt-logs-engine` projects.

For now, continue to use `project.foreman-*` in the index pattern, and for testing set your index name to be `project.foreman-main.` and create an OpenShift project `foreman-main`

Comment (lzap) [reply templates/foreman/template.yml:6]: Let me change this, so the end result in rsyslog.conf would be:

    searchIndex="project.foreman-logs.2018.11.07"

Is that correct?

Comment (richm) [reply templates/foreman/template.yml:6]: yes, looks good - so the index_pattern can remain `project.foreman-*`

Comment (lzap) [reply templates/foreman/template.yml:6]: Ok thanks!

**Comment by lzap** [comment]
Ok I have tested this and it works fine, one change I amended was:

```diff
diff --git a/namespaces/foreman.yml b/namespaces/foreman.yml
index cf8c3e7..f42ee24 100644
--- a/namespaces/foreman.yml
+++ b/namespaces/foreman.yml
@@ -59,6 +59,7 @@ namespace:
     fields:
       - name: message
         type: text
+        doc_values: false
         description: >
           Exception message
           
@@ -69,6 +70,7 @@ namespace:
           
       - name: backtrace
         type: text
+        doc_values: false
         description: >
           Full exception backtrace (multiline text).
           
diff --git a/templates/foreman/template.yml b/templates/foreman/template.yml
index 14a5abe..4c88b92 100644
--- a/templates/foreman/template.yml
+++ b/templates/foreman/template.yml
@@ -5,10 +5,6 @@ elasticsearch_template:
   name: org.foreman.viaq-cdm
   index_pattern: "project.foreman-*"
   order: 20
-  _source:
-    enabled: false
-  _all:
-    enabled: false
 
 namespaces:
   - foreman.yml
```

Doc_value had to be there, the other one was copy and paste mistake.

**Comment by lzap** [comment]
Is there some generic rule on `index: true` or `false`? Should I look on this as RDBM index? Everytime I want to search or sort entries by given column I want it to be indexed? Or is this fulltext index, therefore only relevant for the message?

**Comment by lzap** [comment]
One more question, I realized that now I send message three times: message, MESSAGE and msg. Which one of these is expected? From the JSON generated template it looks like it should be `message` and all the rest I need to drop.

**Comment by richm** [comment]
> Is there some generic rule on index: true or false?

Yes, specified by the field defaults: https://github.com/ViaQ/elasticsearch-templates/blob/master/namespaces/_default_.yml#L3
```
field_defaults:
  type: keyword
  doc_values: true
  index: true
```
by default, fields are indexed.

> Should I look on this as RDBM index? Everytime I want to search or sort entries by given column I want it to be indexed?

Yes, and used in aggregations.  It isn't exactly like an RDBMS index, but close enough.

> Or is this fulltext index, therefore only relevant for the message?

type: keyword isn't used for fulltext index, use type: text for that.  Typically, fields such as unique ids, names of objects (i.e. in kubernetes, pod names and namespace names) will be keywords, because you typically want to do exact match searches on this field.  Typically the "message" field is type: text and has a fulltext index because it is usually an unstructured blob and is used for "fuzzy" matching searches.

**Comment by richm** [comment]
> One more question, I realized that now I send message three times: message, MESSAGE and msg. Which one of these is expected? From the JSON generated template it looks like it should be message and all the rest I need to drop.

Right, but it is a little bit tricky.
In rhel7, rsyslog variables/fields are case sensitive - so you can have `message` and `MESSAGE` in the same record.  Beginning with rsyslog 8.30 (fedora, rhel8), field names are case insensitive - so if you delete `MESSAGE` that may delete `message`.  This is primarily a problem with the imjournal fields which are UPPER CASE, and some have names which conflict with the common data model.

Here is some code that attempts to work with both versions of rsyslog: https://github.com/openshift/origin-aggregated-logging/blob/master/hack/testing/rsyslog/viaq_formatting.conf#L162

**Comment by lzap** [comment]
Ok so I am removing `index: true` from all fields since it is the default and we are good to go. This is the final version we will use for the first cut.

File: templates/foreman/template.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  order: 20
+
+namespaces:
+  - foreman.yml
+  - systemd.yml
```
Comment (richm) [inline templates/foreman/template.yml:11]: add `- pipeline_metadata.yml` here - see https://github.com/ViaQ/elasticsearch-templates/blob/master/namespaces/pipeline_metadata.yml for some optional fields you can add to your log records which can help with tracking where and how the records came.

**Comment by lzap** [comment]
Ok thanks!

**Comment by lzap** [comment]
One more rebase, I dropped template!context field due to refactoring. What happens if an older version of Satellite sends this fields but its not present in the index? Is that an error, field ignored or added anyway?

**Comment by richm** [comment]
> What happens if an older version of Satellite sends this fields but its not present in the index? Is that an error, field ignored or added anyway?

It is added anyway, due to the dynamic field settings, as type "text" i.e. not a keyword, a full text field, and it will be indexed.

**Review by richm — APPROVED** [review]
/lgtm

**Comment by lzap** [comment]
Thank you just in time! Cheers.

**Pull request merged**